lib/firenest/topology/erlang.ex

```diff
@@ -133,112 +98,327 @@ defmodule Firenest.Topology.Erlang do
+          put_in(state.nodes[node], {id, node_ref, Map.delete(remote_names, name)})
+
+        # We either have a mismatched or an unknown ID because the
+        # PONG message has not been processed yet.
+        _ ->
+          state
+      end
+
     {:noreply, state}
   end
 
-  def handle_info({:ping, other_id, pid}, %{id: id} = state) do
-    pong(pid, id)
-    {:noreply, add_node(state, other_id, pid)}
+  # This message comes from :net_kernel.monitor_nodes/2. Note it is not
+  # guaranteed that we will receive a nodedown before a nodeup with the
+  # same name. More info: http://erlang.org/pipermail/erlang-questions/2016-November/090795.html
+  def handle_info({:nodeup, node, _}, state) do
+    %{topology: topology, id: id, monitors: monitors} = state
+    ping(node, topology, id, Map.to_list(monitors))
+    {:noreply, state}
   end
 
-  def handle_info({:pong, other_id, pid}, state) do
-    {:noreply, add_node(state, other_id, pid)}
+  # Sent by :net_kernel.monitor_nodes/2 on node down. We don't worry
+  # about it because we already monitor the topology process and the
+  # monitor message is guaranteed to be delivered before nodedown.
+  def handle_info({:nodedown, _, _}, state) do
+    {:noreply, state}
   end
 
-  def handle_info({:DOWN, ref, _, pid, _}, state) when Kernel.node(pid) != Kernel.node() do
-    {:noreply, delete_node(state, pid, ref)}
+  # If two nodes come up at the same time, ping may be received twice.
+  # So we need to make sure to handle two pings/pongs.
+  def handle_info({:ping, other_id, pid, monitors}, %{id: id} = state) do
+    pong(pid, id, Map.to_list(state.monitors))
+    {:noreply, add_node(state, other_id, pid, monitors)}
   end
 
-  def handle_info({:DOWN, ref, _, _, _}, state) do
-    {_, state} = pop_in(state.subscribers[ref])
-    {:noreply, state}
+  def handle_info({:pong, other_id, pid, monitors}, state) do
+    {:noreply, add_node(state, other_id, pid, monitors)}
   end
 
-  defp ping(node, topology, id) do
-    send({topology, node}, {:ping, id, self()})
+  ## Helpers
+
+  defp ping(node, topology, id, monitors) when is_list(monitors) do
+    send({topology, node}, {:ping, id, self(), monitors})
   end
 
-  defp pong(pid, id) do
-    send(pid, {:pong, id, self()})
+  defp pong(pid, id, monitors) when is_list(monitors) do
+    send(pid, {:pong, id, self(), monitors})
   end
 
   defp id() do
-    {:crypto.strong_rand_bytes(12), System.system_time()}
+    {:crypto.strong_rand_bytes(4), System.system_time()}
   end
 
-  defp add_node(%{nodes: nodes} = state, id, pid) do
+  defp add_node(%{nodes: nodes} = state, id, pid, monitors) do
     node = Kernel.node(pid)
+    new_remote_names = for {ref, name} <- monitors, do: {name, ref}, into: %{}
+
     case nodes do
-      %{^node => {^id, _}} ->
-        state
+      %{^node => {^id, ref, old_remote_names}} ->
+        :ok = diff_monitors(state, node, id, old_remote_names, monitors)
+        put_in(state.nodes[node], {id, ref, new_remote_names})
```

Comment:
This is the only `case` branch that doesn't notify the subscribers. In that case, the caller of `connect/2` might be left hanging and timeout even though the `{:nodeup, node}` message has arrived. I'm not sure if this scenario is possible to happen due to the order of delivery messages from intermittent nodes, though.

Reply:
In this case, the node is already up, because we are matching on the same node+id pair. So the node is already in `node in nodes` or a node up message was already delivered. So all we need to do is to update the list of remote names.

This happens because the ping/pong messages may be exchanged more than once.